### ISSUE TITLE
ResourceSafetySpec / resource allocation (3) fails

### ISSUE THREAD

**Issue by mpilquist** [functional-streams-for-scala/fs2#591]
The "asynchronous resource allocation (3)" test intermittently either fails or hangs. Opening an issue to track the failures.

```
[info] - asynchronous resource allocation (3) *** FAILED *** (1 second, 455 milliseconds)
[info]   TestFailedException was thrown during property evaluation.
[info]     Message: -1 was not equal to 0
[info]     Location: (ResourceSafetySpec.scala:119)
[info]     Occurred when passed generated values (
[info]       arg0 = PureStream(unchunked,fs2.Stream$$anon$1@71b5aace),
[info]       arg1 = true,
[info]       arg2 = Failure(failure-in-pure-code,fs2.Stream$$anon$1@34878bac),
[info]       arg3 = SmallPositive(1)
[info]     )
```

**Comment by mpilquist** [comment]
Travis hangs often on the "merge/join consistency" test in `ConcurrentSpec`. E.g. https://travis-ci.org/functional-streams-for-scala/fs2/jobs/126896689

**Comment by mpilquist** [comment]
Example travis hang in `asynchronous resource allocation (3)` - https://travis-ci.org/functional-streams-for-scala/fs2/jobs/126896679

**Comment by mpilquist** [comment]
Test hang in `asynchronous resource allocation (3)`:
```
[info]   Occurred when passed generated values (
[info]     arg0 = PureStream(right-associated,fs2.Stream$$anon$1@12210359),
[info]     arg1 = false,
[info]     arg2 = Failure(failure-mid-effect,fs2.Stream$$anon$1@56122253),
[info]     arg3 = SmallPositive(13)
[info]   )
```

**Comment by mpilquist** [comment]
```
[info] - asynchronous resource allocation (3) *** FAILED *** (762 milliseconds)
[info]   TestFailedException was thrown during property evaluation.
[info]     Message: -2 was not equal to 0
[info]     Location: (ResourceSafetySpec.scala:120)
[info]     Occurred when passed generated values (
[info]       arg0 = PureStream(randomly-chunked,fs2.Stream$$anon$1@11afd1f5),
[info]       arg1 = true,
[info]       arg2 = Failure(failure-in-async-code,fs2.Stream$$anon$1@197daf1f),
[info]       arg3 = SmallPositive(5)
[info]     )
```

**Comment by mpilquist** [comment]
I'm pretty sure this is caused by the call to `clear` here: https://github.com/functional-streams-for-scala/fs2/blob/e1b200218fd737956a94fdf73cf5bcb71afcd2fe/core/src/main/scala/fs2/concurrent.scala#L64

If `openLeases.count == maxOpen`, the call to `clear` results in an `IllegalArgumentException` which has the effect of not waiting for inner streams to finish.

**Comment by mpilquist** [comment]
I think the above is the cause of the tests that fail with `-1 was not equal to 0`. I think the cause of tests hanging is the call to `openLeases.decrementBy(maxOpen - m)` being off by 1 due to a race with a blocked `openLeases.decrement`.

**Comment by mpilquist** [comment]
Fixed in #603
